include/cooperative_perception/angle.hpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2022 Leidos
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef COOPERATIVE_PERCEPTION_ANGLE_HPP
+#define COOPERATIVE_PERCEPTION_ANGLE_HPP
+
+#include <complex>
+#include <math.h>
+#include <units.h>
+#include <cooperative_perception/utils.hpp>
+
+namespace cooperative_perception
+{
+/**
+ * @brief Angle type for representing and manipulating angles
+ */
+class Angle
+{
+private:
+  std::complex<double> value_{};
+
+public:
+  Angle(units::angle::radian_t angle_val) noexcept : value_{ units::math::cos(angle_val), units::math::sin(angle_val) }
+  {
+  }
+
+  auto set_angle(units::angle::radian_t angle_val) noexcept -> void
+  {
+    value_ = std::complex<double>{ units::math::cos(angle_val), units::math::sin(angle_val) };
+  }
+
+  auto get_angle() const noexcept -> units::angle::radian_t
+  {
+    auto phase_angle = std::arg(value_);
+    if (phase_angle < 0.0)
+    {
+      phase_angle += 2.0 * M_PI;
+    }
+    return units::angle::radian_t{ phase_angle };
+  }
+};
+
+inline auto operator+=(Angle& lhs, const Angle& rhs) noexcept -> Angle&
+{
+  auto angle_sum = lhs.get_angle() + rhs.get_angle();
+  lhs.set_angle(angle_sum);
+  return lhs;
+}
+
+inline auto operator-=(Angle& lhs, const Angle& rhs) noexcept -> Angle&
+{
+  auto angle_sub = lhs.get_angle() - rhs.get_angle();
+  lhs.set_angle(angle_sub);
+  return lhs;
+}
+
+inline auto operator+(Angle lhs, const Angle& rhs) noexcept -> Angle
+{
+  lhs += rhs;
+  return lhs;
+}
+
+inline auto operator-(Angle lhs, const Angle& rhs) noexcept -> Angle
+{
+  lhs -= rhs;
+  return lhs;
+}
+
+inline auto operator*=(Angle& lhs, double rhs) noexcept -> Angle&
+{
+  lhs.set_angle(rhs * lhs.get_angle());
+  return lhs;
+}
+
+inline auto operator*(Angle lhs, double rhs) noexcept -> Angle
+{
+  lhs *= rhs;
+  return lhs;
+}
+
+inline auto operator*(double lhs, Angle rhs) noexcept -> Angle
+{
+  rhs *= lhs;
+  return rhs;
+}
+
+inline auto operator/=(Angle& lhs, double rhs) noexcept -> Angle&
+{
+  lhs.set_angle(lhs.get_angle() / rhs);
+  return lhs;
+}
+
+inline auto operator/(Angle lhs, double rhs) -> Angle
+{
+  lhs /= rhs;
+  return lhs;
+}
+
+inline auto operator==(Angle lhs, Angle rhs) -> bool
+{
+  return cooperative_perception::utils::almostEqual(units::unit_cast<double>(lhs.get_angle()),
+                                                    units::unit_cast<double>(rhs.get_angle()));
+}
```

Comment:
This function breaks the expected semantics for exact-equality and should be re-implemented. Exactly-equality for floating point types is rarely used, but it is always good to be consistent. It would be better to make this a non-member friend function and compare the underlying `std::complex` objects directly because `std::complex` implements `operator==` already. Consider something like:
```cpp
class Angle
{
private:
    // ...
public:
    friend auto operator==(const Angle& lhs, const Angle& rhs) -> bool
    {
        return lhs.value_ == rhs.value_;
    }
};
```

Almost-equality can be achieved with an `almostEqual` overload like this:
```cpp
namespace utils {  // this is nested inside the `cooperative_perception` namespace
    almostEqual(const Angle& lhs, const Angle& rhs)
    {
        return lhs.get_angle() == rhs.get_angle();
    }
}
```
**Note:** The `almostEqual` implementation compares almost-equality between two `units::angle::radian_t` types. The `operator==` overload in the `units` library does not follow the exact-equality semantics. Instead, it checks for almost-equality. See the relevant discussion [here](https://github.com/nholthaus/units/issues/118).

Reply:
To clarify, your example implementation looks like `operator==` is a member. I thought members were always within the `class` declaration section.

Reply:
Functions with the `friend` declaration are non-member functions (see [here](https://en.cppreference.com/w/cpp/language/friend)).

The declaration must be within the class body, but the definition can be elsewhere. I think it is better style to put the definition in the class body because the function implementation accesses private member data. You could alternatively do this (but I don't think it looks as nice):

```cpp
class Angle
{
private:
    // ...
public:
    friend auto operator==(const Angle& lhs, const Angle& rhs) -> bool;
};

auto operator==(const Angle& lhs, const Angle& rhs) -> bool
{
    return lhs.value_ == rhs.value_;
}
```

Reply:
I see - that was a subtlety I missed. I will go with the implementation within the `class` structure as it makes it easier to read.